File: Knapsack/SimpleKnapsackModel.py
from Domain.Models.Schedule import Schedule
from Domain.Models.Enums.Grade import Grade


class SimpleKnapsackModel:
    def __init__(self, schedule):
        self.__schedule = Schedule(schedule.shifts, schedule.nurses)
        # Amount of nurses required to fill all night shifts
        self.E = sum(shift.coverRequirements[Grade.THREE] for shift in schedule.shifts if shift.isNightShift())
        # Amount of nurses required to fill all day shifts 
        self.D = sum(shift.coverRequirements[Grade.THREE] for shift in schedule.shifts if not shift.isNightShift())

        # Target, optimal value for knapsack problem
        self.Z = 0

        

        print("Schedule requires: " + str(self.e) + " nurses to work all night shifts")
        print("Schedule requires: " + str(self.d) + " nurses to work all day shifts")

    def searchTree(self):
        return self.__schedule.nurses

File: Knapsack/SimpleKnapsackModel.cs
using Domain.Models;
using Domain.Models.Enums;

namespace Knapsack;

public class SimpleKnapsackModel : IKnapsackModel
{
    // Total number of required night shifts
    private readonly int E;
    // Total number of required day shifts 
    private readonly int D;
    // The schedule containing the shifts to be covered
    private Schedule Schedule;

    public SimpleKnapsackModel(Schedule schedule)
    {
        Schedule = schedule;

        // Måske har jeg forstået den her del forkert fra artiklen

        E = Schedule.Shifts
            .Where(x => x.IsNightShift())
            .Select(x => x.CoverRequirements[Grade.THREE])
            .Sum();

        D = Schedule.Shifts
            .Where(x => !x.IsNightShift())
            .Select(x => x.CoverRequirements[Grade.THREE])
            .Sum();

        Console.WriteLine("Schedule requires: " + E + " nurses to work all night shifts");
        Console.WriteLine("Schedule requires: " + D + " nurses to work all day shifts");
    }

    public List<Nurse> SearchTree()
    {
        // Just return the same nurses from input
        return Schedule.Nurses;
    }
}
